{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "86ad2bb7-7fce-4ccb-b2ba-c036f0eda46b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting biopython\n",
      "  Downloading biopython-1.84-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m2.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m0:01\u001b[0m:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: numpy in /usr/lib/python3/dist-packages (from biopython) (1.24.2)\n",
      "Installing collected packages: biopython\n",
      "Successfully installed biopython-1.84\n"
     ]
    }
   ],
   "source": [
    "#some of the preliminary commands for assembly before analysis\n",
    "#seqtk sample -s 828 oil_R1.fastq 5000000 > pesub1.fastq\n",
    "#seqtk sample -s 828 oil_R2.fastq 5000000 > pesub2.fastq\n",
    "#seqtk sample -s 828 oilMP_S4_L001_R1_001.fastq 1500000 > mpsub1.fastq\n",
    "#seqtk sample -s 828 oilMP_S4_L001_R2_001.fastq 1500000 > mpsub2.fastq\n",
    "#cat mpsub1.fastq\n",
    "#mkdir seqtk_out\n",
    "#mv mpsub* seqtk_out/\n",
    "#mv pesub* seqtk_out/\n",
    "#cd seqtk_out/\n",
    "#multiqc .\n",
    "#fastqc mpsub1.fastq \n",
    "#fastqc mpsub2.fastq && fastqc pesub1.fastq && fastqc pesub2.fastq \n",
    "#platanus_trim -i pesub1.fastq && platanus_trim -i pesub2.fastq \n",
    "#platanus_trim -i pesub1.fastq \n",
    "#echo \"pesub1.fastq\" > pe.fofn\n",
    "#echo \"pesub2.fastq\" > pe.fofn\n",
    "#platanus_trim -i pe.fofn\n",
    "#platanus_internal_trim -i mp.fofn \n",
    "#fastqc mpsub1.fastq.int_trimmed && fastqc mpsub2.fastq.int_trimmed && fastqc pesub1.fastq.trimmed && fastqc pesub2.fastq.trimmed \n",
    "multiqc .\n",
    "#platanus assemble -o platanus -f *trimmed\n",
    "#platanus scaffold -o platanus -c platanus_contig.fa -b platanus_contigBubble.fa -IP1 *.trimmed -OP2 .int_trimmed\n",
    "#platanus scaffold -o platanus -c platanus_contig.fa -b platanus_contigBubble.fa -IP1 pesub1.fastq.trimmed pesub2.fastq.trimmed -OP2 mpsub1.fastq.int_trimmed mpsub2.fastq.int_trimmed\n",
    "#platanus gap_close -o platanus -c platanus_scaffold.fa -IP1 ../trimmed_reads/pesub1.fastq.trimmed ../trimmed_reads/pesub2.fastq.trimmed -OP2 ../trimmed_reads/mpsub1.fastq.int_trimmed ../trimmed_reads/mpsub2.fastq.int_trimmed\n",
    "\n",
    "!pip install biopython\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73f14f0f-6bfd-4608-8357-c8e3b88d2ee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "def analyze_contigs(file_path):\n",
    "    # read contig file\n",
    "    contigs = list(SeqIO.parse(file_path, \"fasta\"))\n",
    "    \n",
    "    # count total contig number\n",
    "    total_contigs = len(contigs)\n",
    "    \n",
    "    # sum length of all contigs\n",
    "    total_length = sum(len(record.seq) for record in contigs)\n",
    "    \n",
    "    # find length of longest contig\n",
    "    max_length = max(len(record.seq) for record in contigs)\n",
    "    \n",
    "    # calculate N50\n",
    "    sorted_lengths = sorted([len(record.seq) for record in contigs], reverse=True)\n",
    "    cumulative_sum = 0\n",
    "    for length in sorted_lengths:\n",
    "        cumulative_sum += length\n",
    "        if cumulative_sum >= total_length / 2:\n",
    "            n50 = length\n",
    "            break\n",
    "    \n",
    "    return {\n",
    "        \"total number of contigs\": total_contigs,\n",
    "        \"total length of contigs\": total_length,\n",
    "        \"length of longest contig\": max_length,\n",
    "        \"N50\": n50\n",
    "    }\n",
    "\n",
    "contigs_file = \"platanus_contig.fasta\"\n",
    "\n",
    "# contig analysis\n",
    "results = analyze_contigs(contigs_file)\n",
    "print(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "721db220-7d44-45d5-a82d-c5724f88fa88",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_gaps(sequence):\n",
    "    gaps = []\n",
    "    current_gap_start = None\n",
    "    \n",
    "    for i, char in enumerate(sequence):\n",
    "        if char == 'N':\n",
    "            if current_gap_start is None:\n",
    "                current_gap_start = i\n",
    "        elif current_gap_start is not None:\n",
    "            gaps.append((current_gap_start, i - 1))\n",
    "            current_gap_start = None\n",
    "    \n",
    "    if current_gap_start is not None:\n",
    "        gaps.append((current_gap_start, len(sequence) - 1))\n",
    "    \n",
    "    return gaps\n",
    "    \n",
    "def analyze_scaffolds(file_scaffolds):\n",
    "    # read files with FASTA-scaffolds and scaffolds with closed gaps \n",
    "    scaffolds = list(SeqIO.parse(file_scaffolds, \"fasta\"))\n",
    "    #closed_gaps = list(SeqIO.parse(file_closed_gaps, \"fasta\"))\n",
    "    \n",
    "    # count total scaffold number\n",
    "    total_scaffolds = len(scaffolds)\n",
    "    \n",
    "    # sum length of all scaffolds\n",
    "    total_length = sum(len(record.seq) for record in scaffolds)\n",
    "    \n",
    "    # length of longest scaffold\n",
    "    max_length = max(len(record.seq) for record in scaffolds)\n",
    "    \n",
    "    # N50\n",
    "    sorted_lengths = sorted([len(record.seq) for record in scaffolds], reverse=True)\n",
    "    cumulative_sum = 0\n",
    "    for length in sorted_lengths:\n",
    "        cumulative_sum += length\n",
    "        if cumulative_sum >= total_length / 2:\n",
    "            n50 = length\n",
    "            break\n",
    "            \n",
    "    longest_scaffold = next(record for record in scaffolds if len(record.seq) == max_length)\n",
    "    num_gaps = len(count_gaps(longest_scaffold.seq))\n",
    "    gap_lengths = sum([gap[1] - gap[0] + 1 for gap in count_gaps(longest_scaffold.seq)])\n",
    "    \n",
    "    return {\n",
    "        \"total number of scaffolds\": total_scaffolds,\n",
    "        \"total scaffolds length\": total_length,\n",
    "        \"length of longest scaffold\": max_length,\n",
    "        \"N50\": n50,\n",
    "        \"number of gaps in longest scaffold\": num_gaps,\n",
    "        \"total gap length in longest scaffold\": gap_lengths\n",
    "    }\n",
    "\n",
    "file_scaffolds = \"platanus_scaffold.fasta\"\n",
    "#file_closed_gaps = \"platanus_gapClosed.fa\"\n",
    "\n",
    "results = analyze_scaffolds(file_scaffolds)\n",
    "print(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624bcf34-ce07-4ab9-ac17-6dae3cdfbf7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_scaffolds(file_scaffolds):\n",
    "    #scaffolds = list(SeqIO.parse(file_scaffolds, \"fasta\"))\n",
    "    closed_gaps = list(SeqIO.parse(file_closed_gaps, \"fasta\"))\n",
    "    \n",
    "    total_scaffolds = len(closed_gaps)\n",
    "    \n",
    "    total_length = sum(len(record.seq) for record in closed_gaps)\n",
    "    \n",
    "    max_length = max(len(record.seq) for record in closed_gaps)\n",
    "    \n",
    "    sorted_lengths = sorted([len(record.seq) for record in closed_gaps], reverse=True)\n",
    "    cumulative_sum = 0\n",
    "    for length in sorted_lengths:\n",
    "        cumulative_sum += length\n",
    "        if cumulative_sum >= total_length / 2:\n",
    "            n50 = length\n",
    "            break\n",
    "    \n",
    "    longest_scaffold = next(record for record in closed_gaps if len(record.seq) == max_length)\n",
    "    with open('longest.fasta', 'w') as f:\n",
    "        SeqIO.write(longest_scaffold, f, 'fasta')\n",
    "    num_gaps = len(count_gaps(longest_scaffold.seq))\n",
    "    gap_lengths = sum([gap[1] - gap[0] + 1 for gap in count_gaps(longest_scaffold.seq)])\n",
    "    \n",
    "    return {\n",
    "        \"total scaffold number\": total_scaffolds,\n",
    "        \"total scaffold length\": total_length,\n",
    "        \"length of longest scaffold\": max_length,\n",
    "        \"N50\": n50,\n",
    "        \"number of gaps in longest scaffold\": num_gaps,\n",
    "        \"total gap length in longest scaffold\": gap_lengths,\n",
    "        \"longest scaffold\": longest_scaffold\n",
    "    }\n",
    "\n",
    "#file_scaffolds = \"platanus_scaffold.fa\"\n",
    "file_closed_gaps = \"platanus_gapClosed.fasta\"\n",
    "\n",
    "results = analyze_scaffolds(file_closed_gaps)\n",
    "print(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0bb7a47-e691-44c3-b5df-b344f2f45f79",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
